src/kalshi.rs

```diff
@@ -0,0 +1,377 @@
+            // Remove the surrounding "" marks
+            let mut replacement_value = rulebook[key].to_string();
+            replacement_value = replacement_value[1..replacement_value.len()-1].to_string();
+
+            // We've seen instances with and without spaces
+            return_string = return_string.replace(&format!("||{}||", key), &replacement_value);
+            return_string = return_string.replace(&format!("|| {} ||", key), &replacement_value);
+        }
+        return return_string
+
+    }
+
+    pub fn get_resolution_sources_markdown(&self) -> String {
+        let sources = self
+            .event
+            .settlement_sources
+            .iter()
+            .map(|source| format!("<{}>", source.url.clone()))
+            .collect::<Vec<String>>();
+
+        // Return "" if there are none
+        if sources.is_empty() {
+            return "".to_string();
+        }
+        format!("\n\n\n**Resolution sources**\n\n{}", sources.join(", "))
+    }
+
+    pub fn get_binary_resolution(&self) -> Result<Option<BinaryResolution>> {
+        if self.is_resolved() {
+            match self.get_market().result {
+                Some(KalshiResult::Yes) => Ok(Some(BinaryResolution::Yes)),
+                Some(KalshiResult::No) => Ok(Some(BinaryResolution::No)),
+                Some(KalshiResult::StillOpen) => Ok(None),
+                None => Ok(None),
+            }
+        } else {
+            Ok(None)
+        }
+    }
+}
+
+impl TryInto<Question> for &KalshiQuestion {
+    type Error = anyhow::Error;
+
+    fn try_into(self) -> Result<Question> {
+        Ok(Question {
+            source: QuestionSource::Kalshi,
+            source_url: self.full_url(),
+            source_id: self.id.clone(),
+            question: self.get_market().title.clone(),
+            criteria: Some(self.get_criteria_and_sources()),
+            end_date: self.get_market().expiration_date,
+        })
+    }
+}
+
+#[derive(Deserialize, Debug, Clone)]
+pub struct KalshiQuestionsResponse {
+    pub events: Vec<Event>,
+}
+
+#[derive(Deserialize, Debug, Clone)]
+pub struct KalshiQuestion {
+    pub event: Event,
+    #[serde(skip)]
+    pub id: String,
+}
+
+#[derive(Deserialize, Debug, Clone)]
+#[serde(rename_all = "snake_case")]
+pub struct Event {
+    pub series_ticker: String,
+    pub ticker: String,
+    pub markets: Vec<Market>,
+    pub settlement_sources: Vec<SettlementSource>,
+    pub underlying: String,
+}
+
+#[derive(Deserialize, Debug, Clone)]
+pub struct SettlementSource {
+    pub name: String,
+    pub url: String,
+}
+
+#[derive(Deserialize, Debug, Clone)]
+pub struct Market {
+    pub title: String,
+    pub ticker_name: String,
+    pub status: Status,
+    pub open_date: DateTime<Utc>,
+    pub result: Option<KalshiResult>,
+    pub yes_bid: i64,
+    pub yes_ask: i64,
+    pub expiration_date: DateTime<Utc>, // Unsure if we should use close_date, which is earlier
```

Comment:
I think going with the later option is good. We basically never want to close a market that hasn't resolved. We only add the close date for UI/filtering.